### ISSUE TITLE
For a11y, need ability to pass different component than H4 for nonIdealState title, callout title

### ISSUE THREAD

**Issue by bvandrc** [palantir/blueprint#5323]
#### Environment

- __Package version(s)__: 4.3.2
- __Browser and OS versions__: All

#### Feature request

By default, `nonIdealState` and `callout` each make their title an `H4`, and this is not configurable. We need an option to set this to a different component ie `H2`, `div`, `span` (and set custom props for this component as well). Per aXe a11y rules, a page must have descending header levels (i.e., cannot have only an `h1` and `h4`, must have `h1`, `h2`, `h3`, and `h4`, or just `h1` and `h2`, etc..) so many times, this component violates that rule.

**Comment by bvandrc** [comment]
I see now that  `Callout` has a workaround: a user can just pass say, `<div>title string</div>` as the children instead of using `title` if they don't want an `H4` around the title-- `nonIdealState` still has no way to do this though.

**Comment by bvandrc** [comment]
Would advise looking at other usages of header tags as well, as users may not necessarily want them to be header tags (for a11y, or other reasons)-- would probably be better to just use `div` and set the style and classes to be the same as what the header tag previously displayed.

**Comment by adidahiya** [comment]
I hear the request here, and in general I do want to follow a11y rules. How important are these descending header element rules though, really? I don't necessarily want to follow them blindly if they provide no benefit.

If we let users put an `<h2>` in there, Blueprint is still going to style it as an `<h4>` or whatever. So it kind of starts to lose its semantic meaning as an `<h2>`. I don't want to let users set the Callout or NonIdealState title font size to whatever they want; that breaks the typography of the design system.

Is the idea of descending header levels that, without them, screen readers are reading the app content in the wrong order? If so, do you have a concrete example / repro of that problem?

**Comment by bvandrc** [comment]
I'm not entirely sure what the actual impact of this is for screen readers-- We are running aXe a11y testing, which indicated this as a failure. It does indicate the _impact_ as "moderate", not "severe". Here is more info on the rule:

 https://dequeuniversity.com/rules/axe/4.4/heading-order?application=axeAPI

" be sure that you are using the heading markup (h1 through h6's) if and only if you are writing a heading. While applying such markup is a quick way to make text stand out, using it for anything other than headings will make navigating a web page more confusing for users of assistive technology. "

**Comment by adidahiya** [comment]
This may be a case where the heading order a11y rule is overly prescriptive and not helpful. Over my time working with a11y testing and audits I've come to realize that some of these rules are not very well-designed or applicable to modern web applications. Even seemingly reasonable standards like the WCAG color contrast requirements [have their faults](https://tangledweb.xyz/please-stop-using-grey-text-3d3e71acfca8).

**Comment by bvandrc** [comment]
Our team has agreed, we'll opt to ignore the rule. Can close this issue.
